File: dsa-with-subeen-sir/algorithom/inserttion_sort/insertion_sort.py
def insertion_sort(arr):
    size = len(arr)

    # প্রথম element আগেই sorted ধরে নেওয়া হয়
    # তাই index 1 থেকে শুরু
    for i in range(1, size):

        item = arr[i]      # Insert করতে হবে এমন element
        j = i - 1          # Sorted অংশের শেষ index

        # যতক্ষণ আগের element বড় হবে
        # ততক্ষণ এক ঘর ডানে right shift করো
        while j >= 0 and arr[j] > item:
            arr[j + 1] = arr[j] # value 1 position right move
            j -= 1

        # খালি হওয়া position-এ item বসাও
        arr[j + 1] = item


arr = [5, 3, 8, 4, 2]

print("Before:", arr)

insertion_sort(arr)

print("After :", arr)

File: dsa-with-subeen-sir/algorithom/inserttion_sort/insertion_sort.c
#include <stdio.h>

// for new element sort, old array must already sorted
// rules of thumb if less than 100 element and not reverse sorted then use insertion sort
void insertion_sort(int arr[], int size) {
    int i, j, item;
    
    for(i=1; i < size; i++) {
        item = arr[i]; // current value

        // akane item ar jonno ope space find korte hobay
        j = i - 1;
        while(j >= 0 && arr[j] > item) {
            // arr[j] ke tar porer gore (j+1) reke dei
            arr[j+1] = arr[j];
            j = j - 1;
        }

        // a[j+1] holo item ar jonno perfect space
        arr[j+1] = item;
    }
}

int main() {
    int arr[] = {5, 3, 8, 4, 2};
    int size = sizeof(arr) / sizeof(arr[0]);
    printf("Before: ");

    for (int i = 0; i < size; i++)
    {
        printf("%d ", arr[i]);
    }

    insertion_sort(arr, size);

    printf("\nSorted Array: ");
    for (int i = 0; i < size; i++)
    {
        printf("%d ", arr[i]);
    }

    return 0;
}